### ISSUE TITLE
ability to inject mathjax onto cljdoc pages?

### ISSUE THREAD

**Issue by sritchie** [cljdoc/cljdoc#434]
In https://github.com/sicmutils/sicmutils we have a lot of mathematical functions that would benefit from the ability to render LaTeX on our article pages (or in docstrings, now that I think of it!!)

I don't think we need to do anything special, other than all the user to opt some script block like this in to the footer of the page:

```
script>
window.MathJax = {
  loader: {load: ['[tex]/physics']},
  tex: {
    tags: 'ams',
    processEscapes: true,
    inlineMath: {'[+]': [['$', '$']]},
  }
};
</script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js">
</script>
```

This would let forms like $\sin \phi$ render as TeX.

**Comment by lread** [comment]
Hiya @sritchie! Asciidoctor can render equations if the support is enabled.

See: https://asciidoctor.org/docs/user-manual/#activating-stem-support

And it does reach out to MathJax.

When I last stumbled on it, I though "neat!" and I also wondered how it would affect offline docs.

**Comment by sritchie** [comment]
oh, wow, so this will Just Work, maybe?? amazing! I'll give it a try.

**Comment by lread** [comment]
@sritchie, it won't work quite yet because cljdoc hasn't enabled it yet. 
But the feature is available to cljdoc.  
We'd just have to understand the implications of turning it on.

You can get a feel for what your math would look like by previewing locally.  Personally I use [this browser extension](https://github.com/asciidoctor/asciidoctor-browser-extension) to preview locally. It's not how cljdoc renders, but gives you an idea.

So if you try to locally preview this doc:
```asciidoctor
= Stem test
:stem:

stem:[sqrt(4) = 2]

Water (stem:[H_2O]) is a critical component.
```

You should see something like:
![image](https://user-images.githubusercontent.com/967328/102543618-8fa75880-4081-11eb-9628-8651259ef696.png)

Note that GitHub doesn't render this stuff but that doesn't mean cljdoc can't offer a better experience.

**Comment by sritchie** [comment]
Okay, that's great.

I had put together this plugin for mathjax rendering on Github: https://chrome.google.com/webstore/detail/mathjax-3-plugin-for-gith/peoghobgdhejhcmgoppjpjcidngdfkod

which takes care of most of the math in comments (with some breakages, `;;` freaks out the LaTeX renderer). Cljdoc support here would be really fantastic if we end up being able to turn it on!

**Comment by lread** [comment]
Oh that's an interesting way to do what GitHub won't!

> Cljdoc support here would be really fantastic if we end up being able to turn it on!

Yeah I thought it would be cool too, and I don't even have a use for it! 🙂 

@martinklepsch do you have any concerns?  

One thing that might not immediately work is rendering of math for offline docs.  Depending on how many folks rely on offline docs, we could probably work something out.

**Comment by martinklepsch** [comment]
Only concern is that it encourages writing of documentation that won't work well when browsing on GitHub. When implementing the link rewriting I really wanted to ensure that you can keep links as you would normally on GitHub.

This is a different thing though and I think we could probably support Mathjax or similar. Making it work offline might require some extra steps but nothing too crazy. 

So 👍 from my end ☺️

**Comment by lread** [comment]
@sritchie I'm beginning to take a look at this one, hopefully you are still interested?

Before I get too far along, I'd like to make sure I'm about to do something that will work for both cljdoc and you.
Can you point me to one of your adoc files that is rife with formulas?
I did a grep for `:stem:` in sicmutils, but did not find any matches, [maybe you are not using the adoc notation](https://docs.asciidoctor.org/asciidoc/latest/stem/stem/#inline) yet? 

Technical cljdoc notes for myself:
- Asciidoctor will include necessary stem support (MathJax) in an HTML page, but only when rendering an entire doc (with headers and footers); we are rendering an embedable doc (without headers and footers), so we'll have to mimic what Asciidoctor does. Not very difficult, but something to know.
- When Asciidoctor  adds MathJax support when rendering entire docs, it only does so when stem is activated in a doc. I think it makes sense for us to do the same. We don't get access to an AST for adocs while converting, but I think a simple search in adoc content for any line beginning with `:stem:` should suffice.
- The cljdoc rich text render will likely also return what it needs to have included at the end of the HTML page (namely, sometimes MathJax support!). I'll decide what form this takes as I implement, but it should also support offline rendering.

**Comment by sritchie** [comment]
Hey @lread , I am! I don't have any adoc files full of formulas yet because I didn't expect them to render on cljdoc, but I've got a bunch of namespaces rife with math that I'll move into adoc form.

Here's an example, the `differential` namespace: https://github.com/sicmutils/sicmutils/blob/master/src/sicmutils/differential.cljc

Big chunks of this should live in asciidoc, and render like this: https://samritchie.io/dual-numbers-and-automatic-differentiation/

A more outrageous example is the full `sicmutils.numerical.quadrature` package, rendered here with lots of math: https://samritchie.io/functional-numerical-methods/

> I think a simple search in adoc content for any line beginning with :stem: should suffice.

This sounds great! If it will help I'm happy to get some docs converted now for testing.

Thank you for this, @lread! @mk and I are starting a push on a series of "executable textbooks" based on the SICMUtils codebase, and being able to have the API reference express mathematics will be really powerful.

**Comment by lread** [comment]
Thanks @sritchie! I'll take a look at your links.

Just to be clear: cljdoc will only be able to render formulas from `.adoc` files. 
There won't be any support for doing so from `.md` files.
And cljdoc rendered docstrings only support `.md` (CommonMark) so there won't be any support for formulas in docstrings either.

Given all that, still sound enticing?

As for testing that your formulas render, I'm guessing that we'll be fine.
The Asciidoctor STEM support with MathJax seem to do what you are looking for, right?

**Comment by sritchie** [comment]
@lread yes, that's right, I'll port what I need over to adoc, so this is great. Asciidoctor STEM support seems excellent as well. Thanks!

**Comment by lread** [comment]
Great stuff @sritchie, unless otherwise distracted, I should get started on this in the coming week.  
I'll keep you posted.

**Comment by lread** [comment]
I am making progress and have formulas rendering on cljdoc on my dev box.

My current challenge is getting this all working when [docs are downloaded for offline viewing](https://github.com/cljdoc/cljdoc/blob/master/doc/userguide/for-users.md#offline-docs). 
(@martinklepsch people do use and love the download docs feature right?  We still need it, right? Have I asked this before? 🙂).
AsciiDoc uses MathJax v2.7, which is made up of many little pieces.  
MathJax v3 has been out for a good while and seems to offer packaging support which would simplify things for bundling when downloading. I'll try v3 out and see how it behaves.

If offline downloaded doc support is blocking me, I'll likely ship online only support first. 
I can always create a separate issue for supporting MathJax for downloaded docs.

**Comment by martinklepsch** [comment]
> people do use and love the download docs feature right? We still need it, right? Have I asked this before? 🙂

Yes! The Dash integration relies on offline docs and I know quite a few people who make heavy use of that. 

I think some lag for making MathJax work offline could be ok though. Also happy to pair on this if you'd like 🙂

Excited to see this in action, feel free to share some screenshots ☺️

**Comment by lread** [comment]
> > people do use and love the download docs feature right? We still need it, right? Have I asked this before? 🙂
> 
> Yes! The Dash integration relies on offline docs and I know quite a few people who make heavy use of that.

Ok, I'll stop my whining then. 🙂 Good to know it is valued and used.

> I think some lag for making MathJax work offline could be ok though. Also happy to pair on this if you'd like 🙂

Thanks, I think it is more of a mundane than a tricky problem. 
I'll explore options a bit more today.

> Excited to see this in action, feel free to share some screenshots ☺️

Here's a fresh one from my dev box:
![image](https://user-images.githubusercontent.com/967328/112317459-74507880-8c82-11eb-8ae2-911a4436ee2f.png)

**Comment by lread** [comment]
Some more notes for offline support:

On MathJax v3:
- [MathJax v3 does support custom builds via webpack](http://docs.mathjax.org/en/latest/web/webpack.html). This would be convenient for bundling for the offline download. 
- The folks over at Asciidoctor [are exercising caution on upgrading to MathJax v3](https://github.com/asciidoctor/asciidoctor/issues/3449#issuecomment-670912208).
- MathJax v3 might not be a cooked as I imagined, the MathJax team have not ported over all MathJax v2 features yet. A quick attempt with v3, had my asciimath formulas not rendering, there seemed to be a problem with finding the custom AsciiDoc delimiters (maybe?). This might have been a config issue - MathJax config is a bit complex, and it has changed between v2 and v3.

So, although the webpacking is enticing, we'll stick with AsciiDoctor sanctioned and seemingly solid MathJax v2 for now.

On MathJax v2:
- [Other people have attempted to reduce the size of MathJax](https://github.com/mathjax/MathJax-docs/wiki/Guide:-reducing-size-of-a-mathjax-installation/1814429ed1e97bfb7675c0fd400804baa9287249), we'll take advantage of their research for offline downloads.
- Since we are looking at maybe 700kb zipped (we shall see) for MathJax, we'll only include it in an offline bundle when AsciiDoc STEM is explicitly used by a library's adocs.
